src/librustc/ty/instance.rs

```diff
@@ -111,4 +115,192 @@ impl<'a, 'b, 'tcx> Instance<'tcx> {
     pub fn def_id(&self) -> DefId {
         self.def.def_id()
     }
+
+    /// The point where linking happens. Resolve a (def_id, substs)
+    /// pair to an instance.
+    pub fn resolve(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
I wonder if this should be a query. Seems like a handy thing to cache.

Reply:
If we did make it a query, I'd prefer to see the provider moved into `traits/instance` or something.